docs/book/src/developer/providers/contracts/infra-cluster.md

```diff
@@ -0,0 +1,516 @@
+    Status FooClusterStatus `json:"status,omitempty"`
+}
+
+type FooClusterSpec struct {
+    // See other rules for more details about mandatory/optional fields in InfraCluster spec.
+    // Other fields SHOULD be added based on the needs of your provider.
+}
+
+type FooClusterStatus struct {
+    // See other rules for more details about mandatory/optional fields in InfraCluster status.
+    // Other fields SHOULD be added based on the needs of your provider.
+}
+```
+
+For each InfraCluster resource, you MUST also add the corresponding list resource.
+The list resource MUST be named as `<InfraCluster>List`.
+
+```go
+//+kubebuilder:object:root=true
+
+// FooClusterList contains a list of fooclusters.
+type FooaClusterList struct {
+    metav1.TypeMeta `json:",inline"`
+    metav1.ListMeta `json:"metadata,omitempty"`
+    Items           []FooCluster `json:"items"`
+}
+```
+
+### InfraCluster control plane endpoint
+
+Each Cluster needs a control plane endpoint to sit in front of control plane machines.
+Control plane endpoint can be provided in three ways in Cluster API: by the users, by the control plane provider or
+by the infrastructure provider.
+
+In case you are developing an infrastructure provider which is responsible to provide a control plane endpoint for
+each Cluster, the host and port of the generated control plane endpoint SHOULD surface on `spec.controlPlaneEndpoint`
+in the InfraCluster resource.
+
+```go
+type FooClusterSpec struct {
+    // ControlPlaneEndpoint represents the endpoint used to communicate with the control plane.
+    // +optional
+    ControlPlaneEndpoint APIEndpoint `json:"controlPlaneEndpoint"`
+    
+    // See other rules for more details about mandatory/optional fields in InfraCluster spec.
+    // Other fields SHOULD be added based on the needs of your provider.
+}
+
+// APIEndpoint represents a reachable Kubernetes API endpoint.
+type APIEndpoint struct {
+    // The hostname on which the API server is serving.
+    Host string `json:"host"`
+    
+    // The port on which the API server is serving.
+    Port int32 `json:"port"`
+}
+```
+
+Once `spec.controlPlaneEndpoint` is set on the InfraCluster resource and the [InfraCluster initialization completed],
+the Cluster controller will bubble up this info in Cluster's `spec.controlPlaneEndpoint`.
+
+If instead you are developing an infrastructure provider which is NOT responsible to provide a control plane endpoint,
+the implementer should exit reconciliation until it sees Cluster's `spec.controlPlaneEndpoint` populated.
+
+### InfraCluster failure domains
+
+In case you are developing an infrastructure provider which has a notion of failure domains where machines should be
+placed in, the list of available failure domains SHOULD surface on `spec.failureDomains` in the InfraCluster resource.
+
+```go
+type FooClusterStatus struct {
+    // FailureDomains is a list of failure domain objects synced from the infrastructure provider.
```

Comment:
This is status while the text above and below says spec

(also in "InfraCluster Initialization completed" section)

Reply:
good catch, fixed